src/server/scripts/Northrend/Ulduar/Ulduar/boss_thorim.cpp

```diff
@@ -95,7 +904,1269 @@ class boss_thorim : public CreatureScript
+            }
+
+            void DamageTaken(Unit* attacker, uint32& damage) override
+            {
+                if (!attacker->GetAffectingPlayer())
+                    damage = 0;
+            }
+
+            void ExecuteEvent(uint32 eventId) override
+            {
+                switch (eventId)
+                {
+                    case EVENT_PRIMARY_ABILITY:
+                        if (UseAbility(_info->PrimaryAbility))
+                            _events.ScheduleEvent(eventId, urand(15000, 20000));
+                        else
+                            _events.ScheduleEvent(eventId, 1000);
+                        break;
+                    case EVENT_SECONDARY_ABILITY:
+                        if (UseAbility(_info->SecondaryAbility))
+                            _events.ScheduleEvent(eventId, _info->SecondaryAbility == SPELL_SHOOT ? 2000 : urand(4000, 8000));
+                        else
+                            _events.ScheduleEvent(eventId, 1000);
+                        break;
+                    case EVENT_THIRD_ABILITY:
+                        if (UseAbility(_info->ThirdAbility))
+                            _events.ScheduleEvent(eventId, urand(6000, 8000));
+                        else
+                            _events.ScheduleEvent(eventId, 1000);
+                        break;
+                    default:
+                        break;
+                }
+            }
+        };
+
+        CreatureAI* GetAI(Creature* creature) const override
+        {
+            return GetUlduarAI<npc_thorim_pre_phaseAI>(creature);
+        }
+};
+
+class npc_thorim_arena_phase : public CreatureScript
+{
+    public:
+        npc_thorim_arena_phase() : CreatureScript("npc_thorim_arena_phase") { }
+
+        struct npc_thorim_arena_phaseAI : public npc_thorim_trashAI
+        {
+            npc_thorim_arena_phaseAI(Creature* creature) : npc_thorim_trashAI(creature)
+            {
+                switch (_info->Type)
+                {
+                    case DARK_RUNE_CHAMPION:
+                    case DARK_RUNE_WARBRINGER:
+                    case DARK_RUNE_COMMONER:
+                    case DARK_RUNE_EVOKER:
+                        _isInArena = true;
+                        break;
+                    case DARK_RUNE_ACOLYTE:
+                        _isInArena = (_info->Entry == NPC_DARK_RUNE_ACOLYTE_PRE);
+                        break;
+                    default:
+                        _isInArena = false;
+                        break;
+                }
+
+                if (_isInArena)
+                    SetBoundary(&ThorimInArenaBoundaries);
+                else
+                    SetBoundary(&ThorimOutOfArenaBoundaries);
+            }
+
+            bool CanAIAttack(Unit const* who) const override
+            {
+                // don't try to attack players in balcony
+                if (_isInArena && who->GetPositionZ() > 428.0f)
```

Comment:
I tried, funny things happen with ZRange boundary (you need to specify both a min and a max, and I a min only)